### PR TITLE
remove dependency on bluebird methods from tests

### PR DESCRIPTION
**Pull request by maximelkin**
check https://github.com/knex/knex/pull/3650 for older discussions

### CODE REVIEW

**Comment by briandamaged** [comment]
@maximelkin : Thx for splitting this into a separate PR!  I'll have some time to look over it later this afternoon.

File: test/index.js [hunk 1]
```diff
@@ -4,18 +4,11 @@
 require('source-map-support').install();
 
 global.sinon = require('sinon');
+require('./chai-setup');
+global.chai.use(require('sinon-chai'));
```
Comment (briandamaged) [inline test/index.js]: Perhaps this should be moved into `test/chai-setup.js` as well?

File: test/index.js [hunk 2]
```diff
@@ -4,18 +4,11 @@
...
 
-const chai = (global.chai = require('chai'));
-
-chai.use(require('sinon-chai'));
-chai.should();
-
-const bluebird = require('bluebird');
-global.expect = chai.expect;
 global.d = new Date();
```
Comment (briandamaged) [inline test/index.js]: Hmm... having a global variable named `d` seems a bit dangerous since `d` is commonly used as a temporary variable name.  This could lead to some surprises when writing test cases.

(Fixing this might be out of scope for the current PR, though.  But, if we can fix this in 1-2 lines, then it might be worth it).

File: test/index.js [hunk 3]
```diff
@@ -4,18 +4,11 @@
...
 
-bluebird.longStackTraces();
-
```
Comment (briandamaged) [inline test/index.js]: One concern: `bluebird.longStackTraces()` has actually been very helpful in tracking down race conditions in the code!  We might want to figure out a way to optionally enable this feature just for testing purposes.

(The challenge, of course, is that it then also becomes difficult to verify that the code is working properly _without_ relying upon `Bluebird`.  It's like it needs to be a test dependency, but also _not_ be a test dependency at the same time)

cc: @kibertoad

File: test/integration/builder/additional.js [hunk 4]
```diff
@@ -951,6 +908,7 @@ module.exports = function(knex) {
       const knexPrototype = Object.getPrototypeOf(knexDb.client);
       const originalWrappedCancelQueryCall =
         knexPrototype._wrappedCancelQueryCall;
+
       knexPrototype._wrappedCancelQueryCall = (conn) => {
```
Comment (briandamaged) [inline test/integration/builder/additional.js]: Hmm... we should probably be using a `sinon` sandbox here rather than patching the prototype ourselves.  This will help guarantee that the prototype is restored properly once the test case concludes.  (Although this is out of scope for this PR, it might be worth implementing this fix if it's easy to do)

cc: @kibertoad

File: test/integration/builder/additional.js [hunk 5]
```diff
@@ -961,32 +919,27 @@ module.exports = function(knex) {
...
+      try {
+        await expect(
+          getTestQuery().timeout(queryTimeout, { cancel: true })
+        ).to.be.eventually.rejected.and.deep.include({
+          timeout: queryTimeout,
+          name: 'error',
+          message: `After query timeout of ${queryTimeout}ms exceeded, cancelling of query failed.`,
         });
+
+        knexPrototype._wrappedCancelQueryCall = originalWrappedCancelQueryCall;
```
Comment (briandamaged) [inline test/integration/builder/additional.js]: This relates to my "use `sinon` sandboxing" comment from a moment ago.

Notice: if the previous `expect(..)` statement fails, this this line will never be executed.  This means that the prototype will never be restored, which will lead to some really confusing errors in later test cases.

File: test/integration/builder/additional.js [hunk 6]
```diff
@@ -961,32 +919,27 @@ module.exports = function(knex) {
...
-          expect(_.pick(error, 'timeout', 'name', 'message')).to.deep.equal({
-            timeout: queryTimeout,
-            name: 'error',
-            message: `After query timeout of ${queryTimeout}ms exceeded, cancelling of query failed.`,
-          });
-        })
-        .then(() => {
-          knexPrototype._wrappedCancelQueryCall = originalWrappedCancelQueryCall;
-          return getTestQuery().timeout(secondQueryTimeout, { cancel: true });
-        })
```
Comment (briandamaged) [inline test/integration/builder/additional.js]: Interestingly, it looks like the original test case was missing one of these:

```
        .then(function() {
          throw new Error("Shouldn't have gotten here.");
        })
```

Which means: the original test case was potentially overlooking some failures.

File: test/integration/builder/additional.js [hunk 7]
```diff
@@ -961,32 +919,27 @@ module.exports = function(knex) {
...
+
+        await expect(
+          getTestQuery().timeout(secondQueryTimeout, { cancel: true })
+        ).to.be.eventually.rejected.and.deep.include({
+          timeout: secondQueryTimeout,
+          name: 'KnexTimeoutError',
+          message: `Defined query timeout of ${secondQueryTimeout}ms exceeded when running query.`,
+        });
+      } finally {
+        await knexDb.destroy();
```
Comment (briandamaged) [inline test/integration/builder/additional.js]: It looks like there are several lines between:

1. Instantiating the `knexDb` instance, and
2. The beginning of the `try{} finally{}` code

Which means: if a failure occurs anywhere between these steps, then `knexDb` will never be destroyed.  So, we should probably begin the `try{}` statement a bit earlier.  Ex:

```
const knexDb = new Knex(knexConfig);

// Notice: the `try{}` begins immediately after instantiating `knexDb`
try {
  // Notice: this line used to be outside of the `try{}`
  const rawTestQueries = {
      pg: (sleepSeconds) => `SELECT pg_sleep(${sleepSeconds})`,
  };

  // ... more lines ...
} finally {
  await knexDb.destroy();
}
```

File: test/integration/builder/inserts.js [hunk 8]
```diff
@@ -1148,39 +1147,37 @@ module.exports = function(knex) {
...
+          })
+          .then(function() {
+            expect.fail('Should not reach this point');
+          })
+          .catch(function(error) {
+            //Should reach this point before timeout of 10s
+            expect(error.message.toLowerCase()).to.include(
+              'batchinsertduplicatekey'
+            );
+          });
```
Comment (briandamaged) [inline test/integration/builder/inserts.js]: Most of this diff was due to indentation.  Short story: code is establishing the timeout via `chai` instead of `Bluebird` now.

File: test/integration/builder/transaction.js [hunk 9]
```diff
@@ -290,13 +290,14 @@ module.exports = function(knex) {
             expect(count).to.equal(5);
             return knex('test_schema_migrations').count('*');
           })
+          .then(() => expect.fail('should never reach this line'))
```
Comment (briandamaged) [inline test/integration/builder/transaction.js]: Nice fix!

**Comment by kibertoad** [comment]
@briandamaged I agree with your observations regarding tech debt accrued in our tests, we should create a JIRA issue summing up the scope of needed changes.

File: test/integration/builder/transaction.js [hunk 10]
```diff
@@ -489,14 +493,12 @@ module.exports = function(knex) {
...
-        .then(function() {
-          throw new Error('should not get here');
-        })
-        .catch(KnexTimeoutError, function(error) {});
+      try {
+        await db.transaction(() => db.transaction(() => ({})));
+        expect.fail('should not get here');
+      } catch (error) {
+        expect(error).to.be.an.instanceof(KnexTimeoutError);
+      }
```
Comment (briandamaged) [inline test/integration/builder/transaction.js]: FYI: You can probably also do:

```
await expect(
  db.transaction(() => db.transaction(() => ({})))
).to.be.rejectedWith(KnexTimeoutError)
```

File: test/integration/builder/transaction.js [hunk 11]
```diff
@@ -397,8 +401,8 @@ module.exports = function(knex) {
       if (/mssql/i.test(knex.client.driverName)) {
         return Promise.resolve();
       }
-      const first = Bluebird.delay(50);
-      const second = first.then(() => Bluebird.delay(50));
+      const first = delay(50);
+      const second = first.then(() => delay(50));
```
Comment (briandamaged) [inline test/integration/builder/transaction.js]: Your fix here looks good, but this is another test case we should revisit later on.

Looking at the code, I'm not convinced that this test case is sufficient / reliable.  (Especially since we recently discovered that some `Transaction` subclasses were not handling this case properly).  So, we'll need to take another look at this once we finish w/ the Bluebird removal.

File: test/index.js (hunk 1 above)
Comment (maximelkin) [reply test/index.js]: Nope, because this related to sinon too

File: test/index.js (hunk 2 above)
Comment (maximelkin) [reply test/index.js]: note: usages can be found via search /* global.?d */

File: test/index.js (hunk 3 above)
Comment (maximelkin) [reply test/index.js]: Not affects tests directly, so we can just return it back and remove in second pr

File: test/integration/builder/additional.js (hunk 7 above)
Comment (maximelkin) [reply test/integration/builder/additional.js]: As i can see, lines between try block and knex instantiation dont throws, except if
I move new Knex after if

File: test/integration/builder/transaction.js (hunk 10 above)
Comment (maximelkin) [reply test/integration/builder/transaction.js]: done

File: test/integration/builder/additional.js (hunk 7 above)
Comment (briandamaged) [reply test/integration/builder/additional.js]: Given that this is a test case, it's not really a huge concern.  But, generally speaking, it's best to establish "object disposal guarantees" immediately after instantiating an object.  For example, consider the following scenario:

```
const connection = await createConnection();

// But... we haven't established the connection disposal guarantee yet!
doSomethingwith(connection);

// At last, we're setting up the connection disposal guarantee
try {
  // Do stuff...
} finally {
  await connection.dispose();
}
```

Even if `doSomethingwith(connection)` is safe today, it might start throwing exceptions in a future version of the code.  This will then introduce a connection leak into the code, which might prove to be very difficult to find.

File: test/tape/crossdb-compatibility.js [hunk 12]
```diff
@@ -12,97 +13,99 @@ module.exports = function(knex) {
...
+        ]);
+
+        const res = await knex('test_table');
+        t.assert(
+          res.length === 5,
+          'multiple rows with nulls could be added despite of unique constraints'
+        );
+      } finally {
+        t.end();
+      }
```
Comment (briandamaged) [inline test/tape/crossdb-compatibility.js]: The logic appears to be equivalent.

(Side note: this really seems like it should be at least 3 separate test cases.  Given the current design of the test case, the test runner will only document assertions that succeeded.  If an assertion fails, the test runner will just report something like:  "Expected 3 assertions  (Only detected 2)" )

File: test/tape/crossdb-compatibility.js [hunk 13]
```diff
@@ -12,97 +13,99 @@ module.exports = function(knex) {
...
+            t.unique('third');
+            t.unique('fourth');
+          });
+        t.assert(
+          true,
+          'Creating / dropping / creating unique constraint was a success'
+        );
+      } finally {
+        t.end();
+      }
```
Comment (briandamaged) [inline test/tape/crossdb-compatibility.js]: Looks good!  (Another  `.finally(..)` -> `finally{}` replacement)

File: test/unit/knex.js [hunk 14]
```diff
@@ -285,11 +284,10 @@ describe('knex', () => {
 
     const knexWithParams = knex.withUserParams({ userParam: '451' });
 
-    return knexWithParams.transaction((trx) => {
+    return knexWithParams.transaction(async (trx) => {
       expect(trx.userParams).to.deep.equal({
         userParam: '451',
       });
-      return bluebird.resolve();
```
Comment (briandamaged) [inline test/unit/knex.js]: I see.  So, this was to work around the `Transaction` class's special handling for callbacks that return Promises.  (Otherwise, the `Transaction` class would assume that the caller is responsible for committing / rolling back the transaction)

File: test/tape/pool.js [hunk 15]
```diff
@@ -30,51 +30,37 @@ test(`pool evicts dead resources when factory.validate rejects`, (t) => {
...
-    });
+      )
+    );
+
+  const connectionsGroup1 = await allocateConnections();
+  connectionsGroup1.forEach((con) => pool.release(con));
+  connectionsGroup1.forEach((con) => {
+    // fake kill connections
+    con.error = 'connection lost';
+  });
```
Comment (briandamaged) [inline test/tape/pool.js]: It's not important, but it looks like these `.forEach(..)` statements could be combined.

File: test/tape/pool.js [hunk 16]
```diff
@@ -30,51 +30,37 @@ test(`pool evicts dead resources when factory.validate rejects`, (t) => {
...
+  connectionsGroup2.forEach((con) => pool.release(con));
+
+  const connectionsGroup3 = await allocateConnections();
+  connectionsGroup3.forEach((con) =>
+    t.ok(con.id > 4 && con.id < 11, 'Released working connection was used')
+  );
+  connectionsGroup3.forEach((con) => pool.release(con));
+
+  await pool.destroy();
+  t.end();
```
Comment (briandamaged) [inline test/tape/pool.js]: The logic here appears to be equivalent to the logic it is replacing.

File: test/tape/pool.js [hunk 17]
```diff
@@ -101,39 +87,36 @@ test('#823, should not skip pool construction pool config is not defined', funct
...
+        )
+      );
+    } catch (e) {
+      t.fail(
+        `Should have created new connection and execute the query, got : ${e}`
+      );
+    } finally {
+      t.end();
+      await knex.destroy();
+    }
```
Comment (briandamaged) [inline test/tape/pool.js]: The new logic looks like it's equivalent to the old logic.

File: test/tape/transactions.js [hunk 18]
```diff
@@ -318,27 +308,24 @@ module.exports = function(knex) {
...
-          'Second sibling transaction completed'
-        );
       });
+    } finally {
+      t.equal(
+        secondTransactionCompleted,
+        true,
+        'Second sibling transaction completed'
+      );
+    }
```
Comment (briandamaged) [inline test/tape/transactions.js]: `.finally(..)` -> `finally{}`

File: test/tape/transactions.js [hunk 19]
```diff
@@ -352,18 +339,17 @@ module.exports = function(knex) {
...
-      })
-      .finally(function() {
-        return knex('test_table').then(function(results) {
-          t.equal(results.length, 2, 'Parent transaction inserted two records');
-        });
       });
+    } finally {
+      const results = await knex('test_table');
+      t.equal(results.length, 2, 'Parent transaction inserted two records');
+    }
```
Comment (briandamaged) [inline test/tape/transactions.js]: `.finally(..)` -> `finally{}`

File: test/tape/transactions.js [hunk 20]
```diff
@@ -380,19 +366,19 @@ module.exports = function(knex) {
...
-          'Second sibling transaction completed'
-        );
       });
+    } finally {
+      t.equal(
+        secondTransactionCompleted,
+        true,
+        'Second sibling transaction completed'
+      );
+    }
```
Comment (briandamaged) [inline test/tape/transactions.js]: `.finally(..)` -> `finally{}`

File: test/tape/transactions.js [hunk 21]
```diff
@@ -465,20 +449,19 @@ module.exports = function(knex) {
...
+        true,
+        'Second sibling transaction completed'
+      );
+      const results = await knex('test_table');
+      t.equal(results.length, 1, 'Parent transaction inserted one record');
+    }
   });
 
-  test('#625 - streams/transactions', 'postgresql', function(t) {
+  test('#625 - streams/transactions', 'postgresql', (t) => {
```
Comment (briandamaged) [inline test/tape/transactions.js]: Just curious: was there a reason that this needed to be changed into an arrow function?

File: test/index.js (hunk 1 above)
Comment (briandamaged) [reply test/index.js]: It's only "related to sinon" in the sense that it delegates to the sinon library under the hood.  However, it does not actually change the behavior of sinon itself in any way.  (It's a plugin that adds a handful of new functions to chai)

But yeah -- it's not really important for this PR anyway, so we can skip this change for now.

File: test/tape/transactions.js (hunk 21 above)
Comment (kibertoad) [reply test/tape/transactions.js]: In general we are replacing anonymous functions with arrow functions, so there didn't have to be need.

**Comment by kibertoad** [comment]
@briandamaged Yeah, it looks good to me, and you did very through analysis on it already, so I think we are good to go. I created separate issue to track test refactorings.

**Comment by kibertoad** [comment]
@briandamaged Will you do the honours?

**Comment by briandamaged** [comment]
@kibertoad : Done!
